### ISSUE TITLE
Dell Lattitude 5x20/7x20 throttled to ~1500MHz/1800MHz with 2.4.8

### ISSUE THREAD

**Issue by VitaliiSerdiuk** [intel/thermal_daemon#334]
Countinue https://github.com/intel/thermal_daemon/issues/293# as it wasn't properly fixed.

When using google meet video conference + another browser search CPU throttled to 1500 MHz
Latitude 5420
BIOS 1.14.1
Ubuntu - 20.04
Kernel - 5.15.13

**Comment by binboum** [comment]
What version of thermald are you using ? `thermald --version`

Have you compiled version 2.4.8 and installed ? 

I have a Latitude 7490, since the last release, I'm no longer stuck at 400 MHz

**Comment by sebastianha** [comment]
It only affects the 7x20 series.
Also "stuck at 400MHZ" is not correct, with thermald it is stuck at 1800MHz after a short while. The issue is misleading, you have to dig through #293 for all information. But this issue has been closed without reason, so this issue here is only to remind that it has not been solved I think.

**Comment by spamik** [comment]
One funny thing, not sure if it can mean something. I have tried reloading these kernel modules as I have seen in throttled ticket:

rmmod intel_rapl_msr
rmmod processor_thermal_device_pci_legacy
rmmod processor_thermal_device
rmmod processor_thermal_rapl
rmmod intel_rapl_common
rmmod intel_powerclamp

modprobe intel_powerclamp
modprobe intel_rapl_common
modprobe processor_thermal_rapl
modprobe processor_thermal_device
modprobe processor_thermal_device_pci_legacy
modprobe intel_rapl_msr

Behaviour after that is almost same (still fixed to 1800MHz under heavy load with temperatures on 50°C) but two things changed: start of the load is better - I can see for 1-2 seconds frequency over 4GHz and CPU temperature around 90°C (before that I've never reached more than 60). And second strange thing I can hear some clicking noise coming from laptop :-) It happend sometimes - few clicks in a row and after that several tens of seconds or minutes silent...

**Comment by VitaliiSerdiuk** [comment]
Another issue. Thermald service not started after PC restart. I should run it manually

**Comment by sebastianha** [comment]
Dell is collection information: https://www.dell.com/community/Latitude/Latitude-5420-7420-7520-CPU-Throttling-Issue-on-Linux/m-p/8129749/highlight/true#M39458

**Comment by pjssilva** [comment]
It seems also to affect the Lat. 5421 with the i5-11500H  processor. It dips to 800 MHz under full load for some time, and bounces back to the correct 2900 MHz in a continuous cycle.

In my system, a workaround is to change the power profile in the BIOS from "Optimized" to "Ultra performance" (you get louder fans, but the full processor speed). It usually also works well with the "Cool" and "Quiet" profiles. The main problem seems to be the default "Optimized" profile.

Obs: I am using Pop-OS 21.10, kernel 5.15.15 and self compiled 2.4.8 thermald.

**Comment by binboum** [comment]
I think I found a workaround, when I load via USB PD I don't have the problem.

I confirm on the normal charge the problems mentioned.

**Comment by sebastianha** [comment]
What exactly do you mean with Charging via USB or "normal"? My 7320 has only USB-C for charging.

**Comment by binboum** [comment]
> What exactly do you mean with Charging via USB or "normal"? My 7320 has only USB-C for charging.

Like : https://www.dell.com/en-uk/work/shop/dell-uk-65-watt-3-prong-ac-adapter-with-1meter-power-cord/apd/450-aixz/pc-accessories

**Comment by sebastianha** [comment]
Which model do you have? My 7320 has no "normal" power plug.

**Comment by BigYoSpeck** [comment]
I suspect he means charging by the usb-c wall charger vs power delivery like from a dock device

**Comment by sebastianha** [comment]
This is something I have already checked but did not solve the problem for me.

**Comment by BigYoSpeck** [comment]
> This is something I have already checked but did not solve the problem for me.

Dell dock or other brand?

I've not tested this myself yet but I've a Dell TB dock due to arrive soon so I'll feedback

**Comment by sebastianha** [comment]
Both with original AC adapter and with my connected 90w Dell monitor. I also checked BIOS, both are recognized correctly.

**Comment by VitaliiSerdiuk** [comment]
Looks like Dell completely doesn't care about Linux/Open source...

**Comment by JoshuaPK** [comment]
5.16.10-1.el8.elrepo.x86_64 it still exists here as well.  There's one other thing that just came to mind: the fan sensors don't work.  In all of my previous Dell laptops, the system is able to read fan speed from lm_sensors.  In this case the fan speed is not available.  I wonder if the lack of fan speed data is causing thermald to make some assumptions that aren't correct.

**Comment by sebastianha** [comment]
@JoshuaPK for which model are these? I could provide Latitude 7320 with openSuse if needed.

**Comment by pjssilva** [comment]
Here you have the file for a 5421 with i5-11500H. The clock decrease to 800 MHz even with the laptop sitting on top a Zalman laptop base with a fan
[23173225.tar.gz](https://github.com/intel/thermal_daemon/files/8127686/23173225.tar.gz)
.

**Comment by JoshuaPK** [comment]
@sebastianha my apologies.  I have a Latitude 5420 with an i5-1145G7.  I have seen a number of scenarios, but the most frequent are throttling down to 1.5GHz and throttling down to 400MHz.  In my case the throttling goes away when the load decreases.  So, for example, if I try to compress an ISO file with 7za it will throttle, then if I stop 7za it will jump back up to 4GHz.  This is thermald 2.4.8 that I compiled from source, running on Rocky Linux 8.5.

**Comment by spandruvada** [comment]
![issue_334_stress_ng](https://user-images.githubusercontent.com/3802550/155412310-f979ac52-ee78-49c6-b29b-8a2f594ea6fa.png)

**Comment by spandruvada** [comment]
The guaranteed frequency is 1500 MHz on this system. With 100% load system was able to run  about 1000 MHz above guaranteed for 80% of the time. 
What is the expectation? System can't sustain turbo forever.

You can try to manually adjust power and try if you can prevent system for reaching peak turbo and keep above 1500Hz without thermal throttle for test:
Try 
#systemctl disable thermald
Reboot
echo 28000000 > /sys/devices/virtual/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_0_power_limit_uw
echo 28000000 > /sys/devices/virtual/powercap/intel-rapl/intel-rapl:0/constraint_0_power_limit_uw
You can run turbostat
turbostat --show Core,CPU,Busy%,Bzy_MHz,TSC_MHz -o turbostat.out
and check frequencies

**Comment by spandruvada** [comment]
@[pjssilva](https://github.com/pjssilva) Somehow there is no turbostat output in the tar file. Can you see if you have turbostat on your system?

**Comment by VitaliiSerdiuk** [comment]
@spandruvada 
Please could you explain more about 'The guaranteed frequency is 1500 MHz on this system'? For Linux is 1500MHz and for Windows - 2600 MHz, am I correct? Base frequency for Intel Core i5-1145G7 - 2600 MHz. I not understand how it works. So big difference for different OS.

**Comment by spandruvada** [comment]
https://github.com/pjssilva Can you build yourself with the attached?
[0001-Test-patch-to-fail-to-run-adaptive.zip](https://github.com/intel/thermal_daemon/files/8128198/0001-Test-patch-to-fail-to-run-adaptive.zip)

unzip and 
git am 0001-Test-patch-to-fail-to-run-adaptive.patch
then build using make (procedure in README.txt)

Then
systemctl stop thermald
thermald --loglevel=debug --no-daemon --adaptive
Attach the log.
Run some workload also at the same time.

**Comment by spandruvada** [comment]
[@VitaliiSerdiuk 
The guaranteed doesn't change with Linux or Windows. You have a power budget, either you apply more initially and get more short term perf or use it moderately to not drop to 1500 MHz. Linux can't match every conditions in the thermal table as Windows, so Windows may be doing better, but can't say without actually running similar workload. You are running a web/video workload which Windows can manage better based on usage of HW acceleration. So their power usage profile will be different so you may not be dropping to 1500MHz.

**Comment by pjssilva** [comment]
@spandruvada I have installed turbostat and reran the test from the thermal-debug-dump-ubuntu.sh (I run Pop-OS, Ubuntu is the closest).
[23223022.tar.gz](https://github.com/intel/thermal_daemon/files/8129009/23223022.tar.gz)
I will now try to apply the patch you asked in the other message above and will report the results.

**Comment by pjssilva** [comment]
@spandruvada  Now the therlmald log after applying the  0001-Test-patch-to-fail-to-run-adaptive.zip above. Before getting the log I left the system running stress-ng for 10 minutes or so. During the test the system was running for some time at 3600-3700 MHz and then dropping fast to 800 MHz for some time and restarting this cycle. The expected sustained clock for this processor is at least 2400 MHz, so 800 MHz is clearly a problem. 

[thermald.log](https://github.com/intel/thermal_daemon/files/8129179/thermald.log)

**Comment by JoshuaPK** [comment]
This may seem like a silly question, but are some platforms safe to run without thermald?  Using 7za to create an archive of a ~60gb directory.  With thermald running, the system is throttled down to 1.5GHz and stays there, with the core temperature staying steady at around 135F.  If I kill thermald, then the system fluctuates between 2.4 and 3.1GHz (and does not remain below 2.4GHz for any length of time) and the core temperature fluctuates around 140-150F.  I also threw in 8 threads of stress-ng and a YouTube video and the processor still fluctuated around 2.4 but never dipped below 2 for any length of time (temps remained the same aroudn 140).  According to spec the maximum core temperature is 212F.  It appears that in this case the processor is doing a good job of regulating itself.  What am I missing?

**Comment by sebastianha** [comment]
@spandruvada I understand the case with the guaranteed frequency but the problem is, that on my system the fan is not spinning at 100% but the system is throttled down.

In my understanding throttling should only happen when all thermal regulators are maxed out, this means: fan 100%, CPU temperature at 95°C.

Under full load my system is running the fan at ~25-50% and temperature is at 50°C. There is definitely room for more power.

I will deliver the outputs of the scripts as soon I have some spare time left.

**Comment by spandruvada** [comment]
@JoshuaPK The processor has in built in control. But this is about other parts in the system and skin temperature is under spec. But manufacturer may already have made sure out of box. Also the system may already have all the power table configured correctly without thermald. So disable thermald and cold reboot and check, if you still get good performance and decide.

**Comment by spandruvada** [comment]
@sebastianha I am not sure if there is any fan control available. #cat /sys/class/thermal/cooling_device*/type, Do you see any other names other than Processor, LCD,intel_powerclamp? Something like "Fan" or "TFN"

**Comment by sebastianha** [comment]
No:

```
~> cat /sys/class/thermal/cooling_device*/type
Processor
Processor
Processor
Processor
Processor
Processor
Processor
Processor
intel_powerclamp
TCC Offset
```

**Comment by spandruvada** [comment]
So unfortunately we can' t control fans. Do you see same behavior as the plot I attached  above?

**Comment by sebastianha** [comment]
No, I see something like this (manually plotted data):

![Screenshot_20220224_165527](https://user-images.githubusercontent.com/420807/155559766-5c478a99-076e-4af0-acdf-07cad3cd209e.png)

**Comment by sebastianha** [comment]
I also tested with the 0001-Patch:
[thermald-patch0001.txt](https://github.com/intel/thermal_daemon/files/8134252/thermald-patch0001.txt)

**Comment by spandruvada** [comment]
@pjssilva I did some silly mistake. Can you retry with this patch
[0001-Test-patch-to-fail-to-run-adaptive-ver-2.zip](https://github.com/intel/thermal_daemon/files/8134960/0001-Test-patch-to-fail-to-run-adaptive-ver-2.zip)

**Comment by sebastianha** [comment]
[0001-Test-patch-to-fail-to-run-adaptive-ver-2.log](https://github.com/intel/thermal_daemon/files/8135104/0001-Test-patch-to-fail-to-run-adaptive-ver-2.log)

100% load, 50-55°C, no fan, 1800MHz fix the whole time.

**Comment by pjssilva** [comment]
@spandruvada It looks like you have something there! I ran the stress test for 15 minutes and the processor never dipped below 2.9GHz (which is the case clock of the Core i5-11500H configured with high TDP). Take a look at the log below. I will try some other tests, but it looks promising!
[thermald.log](https://github.com/intel/thermal_daemon/files/8135609/thermald.log)

Obs: The dip at the end of the test is because I stopped stress-ng.

**Comment by spandruvada** [comment]
@[sebastianha](https://github.com/sebastianha), you have some other issue. First address issues who get stuck at 800MHz

**Comment by spandruvada** [comment]
@https://github.com/pjssilva
What is the make and model of your system?

**Comment by pjssilva** [comment]
@spandruvada My system is a Dell Latitude 5421, with a Core i5-11500H, 32 GB of RAM, and an Nvidia MX 450 graphics card running in hybrid mode.

**Comment by PhilipGB** [comment]
> Please rub https://github.com/intel/thermal_daemon/blob/master/test/thermal-debug-dump-fedora.sh or https://github.com/intel/thermal_daemon/blob/master/test/thermal-debug-dump-ubuntu.sh. I can check thermal tables first.
[24214023.tar.gz](https://github.com/intel/thermal_daemon/files/8136628/24214023.tar.gz)

Dell Latitude 7320 i7-1185G7 on Ubuntu 20.04.3 running Kernel 5.14.0-1024-oem

**Comment by spandruvada** [comment]
@PhilipGB I don't see any throttling done by thermald.
Can you try this procedure as a root user
- Call systemctl stop thermald (Don't reboot)
- echo 30000000 > /sys/devices/virtual/powercap/intel-rapl-mmio/intel-rapl-mmio\:0/constraint_0_power_limit_uw
Verify by
 cat /sys/devices/virtual/powercap/intel-rapl-mmio/intel-rapl-mmio\:0/constraint_0_power_limit_uw

**Comment by spandruvada** [comment]
If that doesn't address, the next step is also write in addition to above
echo 54000000 > /sys/devices/virtual/powercap/intel-rapl-mmio/intel-rapl-mmio\:0/constraint_1_power_limit_uw

**Comment by PhilipGB** [comment]
Same behaviour. Very briefly clocks to 4.3ghz then settles on 1.8ghz

![image](https://user-images.githubusercontent.com/14321036/155629263-6e99f3cd-41d7-413c-8b29-0a0d8550dd07.png)

I have observed that if I disable the thermald service then reboot the clock speed will float around 2.3ghz with intermittent drops to 400mhz under load

But if it's booted enabled and stopped or has been run and then stopped then even while thermald isn't running the system behaves the same as if it was, settling on 1.8ghz

**Comment by spandruvada** [comment]
Why are you using cpufreq performance. Try to use powersave, it will not hurt performance in most of the cases.

**Comment by spandruvada** [comment]
Also check
cat /sys/bus/pci/devices/0000\:00\:04.0/tcc_offset_degree_celsius
If this is high number write something like "5"

**Comment by spandruvada** [comment]
Also try setting power to 20W.
echo 20000000 > /sys/devices/virtual/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_0_power_limit_uw

I don't see any other special thermal table power value which will change this behavior since thermald is not even throttling here.

**Comment by PhilipGB** [comment]
> Why are you using cpufreq performance. Try to use powersave, it will not hurt performance in most of the cases.

In day to day use that's true and all cores under 100% load both settings almost immediately drop to 1.8ghz anyway but for certain short work loads like running unit tests it does perform about 20% faster with the performance governor

Attempted setting power to 20W and confirmed with cat /sys/devices/virtual/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_0_power_limit_uw being 20000000

But running stress, package-0 power briefly starts at 20W, jumps to 44-50w, then still settles back to 11w at 1.8ghz

**Comment by sebastianha** [comment]
@PhilipGB It seems that we both have the exact same problem (and notebook model). Only difference is, I am running openSuse with kernel 5.16.8-1-default

@spandruvada As we are currently mixing up models and cpus and are seeing different behaviour, would it make sense to split this ticket by creating a single ticket per model/cpu?

**Comment by VitaliiSerdiuk** [comment]
@spandruvada 
Dell Latitude 5420 i5-1145G7 on Ubuntu 20.04.3 running Kernel 5.16.10-051610-generic
[25104402.tar.gz](https://github.com/intel/thermal_daemon/files/8140204/25104402.tar.gz)

**Comment by spandruvada** [comment]
Just to reiterate, the time you can sustain above guaranteed/base frequency is very subjective on workloads on form factor. The idea is to fix so that it drop below this base frequency like 800MHz or 400MHz and give them short term boost . At the same time if some temperature limit is reached, throttle. If you run  some power virus like stress, you will eventually drop to base frequency. This daemon can't fix that.
If you are running some real world use cases like video/Gfx the performance on Windows and Linux will differ depending on how the feature is implemented and HW acceleration support in Linux. For example if Video is decoded in software and played, it will load CPUs more than downloaded to some HW to do the same work. When you compare with windows also check the CPU utilization.

**Comment by spandruvada** [comment]
This analysis for "Latitude 5420" BIOS version:"1.6.0". Since the focus is to match Windows, I started analyzing Windows.

Luckily I got hold of one Dell 5420. I observed that Windows is not updating to any power values from the thermal table. It is possible that something preventing to use these tables, but since I am not a Windows expert, don't quote me on that. If Windows is not using, then there is low confidence that those power values will give any better performance using them.

There was an issue on this platform, where system will run at 400MHz after few seconds, which was addressed before in thermald. But that may have been fixed by BIOS update  later. So that change is still required.

So I created a change which adds a blocked list of platforms which will not use the adaptive tables. With this the performance is matches with Windows, the system stays around 2600MHz.  When you build with the change, you will see in log that "This platform is blocked from using adaptive option Latitude 5420" in the logs. Please make sure that system is powered off and restarted with the new version, not via reboot.
[block_adaptive_cond_latitude_5420.zip](https://github.com/intel/thermal_daemon/files/8144716/block_adaptive_cond_latitude_5420.zip)

**Comment by sebastianha** [comment]
Hi,

I adapted the patch for my 7320, rebooted and started with:

`./thermald --no-daemon --adaptive --loglevel=info`

I see the log entry but the behaviour is still the same as before. Throttling down to 1800MHz, fan not spinning up.

It is a little bit different: I never reached full speed up to 4800MHz and it was a little bit longer at ~2400MHz. But at the end it was still throttled after a few seconds. And still – what I do not understand – very low temperatures and no fan spinning, which should kick in before throttling.

**Comment by sebastianha** [comment]
> This analysis for "Latitude 5420" BIOS version:"1.6.0". Since the focus is to match Windows, I started analyzing Windows.

Just to be sure: As far as I know Windows needs the "Intel Dynamic Tuning Driver" from Dell support to work correctly. But I assume that you have installed this. I think this is where the "magic happens". Some people did also report that they circumvented the throttling problem by rebooting into Linux directly from Windows but I could never reprodruce this (https://www.dell.com/community/Latitude/Latitude-5420-7420-7520-CPU-Throttling-Issue-on-Linux/m-p/8042077/highlight/true#M36369).

**Comment by sebastianha** [comment]
I did and it seems that they did not show any effect.

Do you have explicit commands for me I could try?

**Comment by spandruvada** [comment]
@sebastianha There is a tool in Windows https://www.intel.com/content/www/us/en/developer/articles/tool/power-gadget.html
Suggest to log the power, util, cpu frequency and plot. Start the log before starting a workload and let it run for few min.

**Comment by sebastianha** [comment]
To clean up things I opened a new ticket only for the Latitude 7320 with all my current findings: https://github.com/intel/thermal_daemon/issues/341

**Comment by JoshuaPK** [comment]
I have opened #344 to document information specific to Latitude 5420.

**Comment by spandruvada** [comment]
For issue with 5421, Thermald 2.4.9 has the change.

**Comment by pjssilva** [comment]
I confirm that thermald 2.4.9 fixes the problem in mine Latitude 5421 (core 5-11500H). Great work @spandruvada! Thanks.

**Comment by spandruvada** [comment]
Thanks @pjssilva . Since there are separate issue for Dell 5420 and 7320/7420. This issue is combination of many platform, so better close this.
